azext_iot/iothub/providers/message_route.py

```diff
@@ -0,0 +1,184 @@
+        for route in routes:
+            if route.name.lower() == route_name.lower():
+                return route
+        raise ResourceNotFoundError("No route found.")
+
+    def list(self, source_type: Optional[str] = None):
+        routes = self.hub_resource.properties.routing.routes
+        if source_type:
+            return [route for route in routes if route.source.lower() == source_type.lower()]
+        return routes
+
+    def delete(self, route_name: Optional[str] = None, source_type: Optional[str] = None):
+        routing = self.hub_resource.properties.routing
+        if not route_name and not source_type:
+            routing.routes = []
+        elif route_name:
+            routing.routes = [route for route in routing.routes if route.name.lower() != route_name.lower()]
+        else:
+            routing.routes = [route for route in routing.routes if route.source.lower() != source_type.lower()]
+
+        return self.client.iot_hub_resource.begin_create_or_update(
+            resource_group_name=self.hub_resource.additional_properties['resourcegroup'],
+            resource_name=self.hub_resource.name,
+            iot_hub_description=self.hub_resource,
+            if_match=self.hub_resource.etag
+        )
+
+    def test(
+        self,
+        route_name: Optional[str] = None,
+        source_type: Optional[str] = None,
+        body: Optional[str] = None,
+        app_properties: Optional[str] = None,
+        system_properties: Optional[str] = None
+    ):
+        if app_properties:
+            app_properties = process_json_arg(content=app_properties, argument_name="app_properties")
+        if system_properties:
+            system_properties = process_json_arg(content=system_properties, argument_name="system_properties")
+
+        route_message = {
+            "body": body,
+            "appProperties": app_properties,
+            "systemProperties": system_properties
+        }
+
+        if route_name:
+            route = self.show(route_name)
+            test_route_input = {
+                "message": route_message,
+                "twin": None,
+                "route": route
+            }
+            return self.client.iot_hub_resource.test_route(
+                iot_hub_name=self.hub_resource.name,
+                resource_group_name=self.hub_resource.additional_properties['resourcegroup'],
+                input=test_route_input
+            )
+
+        if source_type:
+            test_all_routes_input = {
+                "routingSource": source_type,
+                "message": route_message,
+                "twin": None
+            }
+            return self.client.iot_hub_resource.test_all_routes(
+                iot_hub_name=self.hub_resource.name,
+                resource_group_name=self.hub_resource.additional_properties['resourcegroup'],
+                input=test_all_routes_input
+            )
+
+        # for all types, need to test all types one by one
```

Comment:
I'm not familiar with previous `route test` functionality.

Could you explain the major differences between the two versions (or at least the previous one)?
This one seems like it:

- Tests a route by name, if provided
- Tests all routes by source_type, if provided
- Tests all source types through all routes if neither are provided, with some logic to dance around the fallback route

Reply:
The biggest change is in the test all source types cause now it actually tests all source types rather than pretend to (and only do device messages)

<img width="692" alt="image" src="https://user-images.githubusercontent.com/73560279/207982425-11f05bf9-d89a-430d-b706-4141a06281fc.png">

old route test - 
![image](https://user-images.githubusercontent.com/73560279/207982455-f6ad7798-7493-494d-9d79-8e1000caca66.png)
only this one passes b/c it is enabled, a true condition, and **is of type DeviceMessages**


new route test - 
![image](https://user-images.githubusercontent.com/73560279/207982651-6e550a94-13a1-4943-ad91-450964ee612c.png)

this grabs things that aren't of type devicemessages

Reply:
added to history